component/attr_cache/attr_cache.go

```diff
@@ -301,79 +253,76 @@ func (ac *AttrCache) renameCachedDirectory(srcDir string, dstDir string, time ti
+		return err
 	}
 
+	srcDir = internal.TruncateDirName(srcDir)
+	dstDir = internal.TruncateDirName(dstDir)
+	ac.renameCachedDirectoryHelper(srcItem, srcDir, dstDir, time)
+
+	ac.updateAncestorsInCloud(srcDir, time)
+	ac.updateAncestorsInCloud(dstDir, time)
+
 	// if there were no cached entries to move, does this directory even exist?
-	if !foundCachedContents && !ac.pathExistsInCache(srcDir) {
+	if srcItem.children == nil && !ac.pathExistsInCache(srcDir) {
 		log.Err("AttrCache::renameCachedDirectory : Source directory %s does not exist.", srcDir)
 		return syscall.ENOENT
 	}
 
-	// record whether the destination directory's parent tree now contains objects
-	if movedObjects {
-		ac.markAncestorsInCloud(dstDir, time)
+	return nil
+}
+
+func (ac *AttrCache) renameCachedDirectoryHelper(srcItem *attrCacheItem, srcDir string, dstDir string, time time.Time) {
+	// we have the source item. look at everything in the source item as reference
+	// for each item in the subtree of itemSrc, create the destination item based on the source item information.
+
+	// 1.) check that source item exists
+	if !srcItem.exists() {
+		return
+	}
+
+	// 2.) take the source name and change it to the destination name
+	dstPath := strings.Replace(srcItem.attr.Path, srcDir, dstDir, 1)
+
+	// 3.) create an attribute using the destination name
+	var dstAttr *internal.ObjAttr
+	if srcItem.attr.IsDir() {
+		dstAttr = internal.CreateObjAttrDir(dstPath)
 	} else {
-		// add the destination directory to our cache
-		dstDir = internal.TruncateDirName(dstDir)
-		dstDirAttr := internal.CreateObjAttrDir(dstDir)
-		dstDirAttrCacheItem := newAttrCacheItem(dstDirAttr, true, time)
-		dstDirAttrCacheItem.markInCloud(false)
-		ac.cacheMap[dstDir] = dstDirAttrCacheItem
+		dstAttr = internal.CreateObjAttr(dstPath, srcItem.attr.Size, srcItem.attr.Mtime)
 	}
 
-	// delete the source directory from our cache
-	ac.deletePath(srcDir, time)
+	// 4.) insert the attribute from previous step into the cacheMap
+	dstItem := ac.cacheMap.insert(dstAttr, srcItem.exists(), srcItem.cachedAt)
 
-	// If this leaves the parent or ancestor directories empty, record that.
-	// Although this involves an unnecessary second traversal through the cache,
-	// because of the code complexity, I think it's worth the readability gained.
-	ac.updateAncestorsInCloud(getParentDir(srcDir), time)
+	// 5.) mark whether the item is in the cloud
+	dstItem.markInCloud(srcItem.isInCloud())
 
-	return nil
+	// 6.) repeat steps 1 - 5 for any children in the current source Item
+	for _, srcChildItm := range srcItem.children {
+		ac.renameCachedDirectoryHelper(srcChildItm, srcDir, dstDir, time)
+	}
+
+	// 7.) if step 1 is true, markAncestorsIncloud(destination name)
+	ac.markAncestorsInCloud(dstPath, time)
+
+	srcItem.markDeleted(time)
 }
 
 func (ac *AttrCache) markAncestorsInCloud(dirPath string, time time.Time) {
 	dirPath = internal.TruncateDirName(dirPath)
```

Comment:
Since `get()` truncates now, there's no need to truncate here anymore. `CreateObjAttrDir()` and `getParentDir()` both truncate already too.